internal/packer_registry/types.bucket.go

```diff
@@ -0,0 +1,288 @@
+	if b == nil {
+		return
+	}
+
+	if _, ok := b.Iteration.builds.m[sourceName]; ok {
+		return
+	}
+
+	build := &Build{
+		ComponentType: sourceName,
+		RunUUID:       b.Iteration.RunUUID,
+		PARtifacts:    make([]PARtifact, 0),
+	}
+	b.Iteration.builds.Lock()
+	b.Iteration.builds.m[sourceName] = build
+	b.Iteration.builds.Unlock()
+}
+
+func (b *Bucket) PublishBuildStatus(ctx context.Context, name string, status models.HashicorpCloudPackerBuildStatus) error {
+	// NOOP
+	if b == nil {
+		return nil
+	}
+
+	// Lets check if we have something already for this build
+	existingBuild, ok := b.Iteration.builds.m[name]
+	if ok && existingBuild.ID != "" {
+		buildInput := &models.HashicorpCloudPackerUpdateBuildRequest{
+			BuildID: existingBuild.ID,
+			Updates: &models.HashicorpCloudPackerBuildUpdates{
+				PackerRunUUID: existingBuild.RunUUID,
+				Status:        &status,
+			},
+		}
+
+		// Possible bug of being able to set DONE with no RunUUID being set.
+		if status == models.HashicorpCloudPackerBuildStatusDONE {
+			images := make([]*models.HashicorpCloudPackerImage, 0, len(existingBuild.PARtifacts))
+			for _, partifact := range existingBuild.PARtifacts {
+				images = append(images, &models.HashicorpCloudPackerImage{ImageID: partifact.ID, Region: partifact.ProviderRegion})
+			}
+			//TODO pass along cloud provider
+			buildInput.Updates.Images = images
+		}
+
+		_, err := UpdateBuild(ctx, b.client, buildInput)
+		if err != nil {
+			return err
+		}
+		b.Iteration.builds.Lock()
+		existingBuild.Status = status
+		b.Iteration.builds.m[name] = existingBuild
+		b.Iteration.builds.Unlock()
+		return nil
+	}
+
+	buildInput := &models.HashicorpCloudPackerCreateBuildRequest{
+		BucketSlug:  b.Slug,
+		Fingerprint: b.Iteration.Fingerprint,
+		Build: &models.HashicorpCloudPackerBuild{
+			ComponentType: name,
+			IterationID:   b.Iteration.ID,
+			PackerRunUUID: b.Iteration.RunUUID,
+			Status:        &status,
+		},
+	}
+
+	id, err := CreateBuild(ctx, b.client, buildInput)
+	if err != nil {
+		return err
+	}
+
+	build := &Build{
+		ID:            id,
+		ComponentType: name,
+		RunUUID:       b.Iteration.RunUUID,
+		Status:        status,
+		PARtifacts:    make([]PARtifact, 0),
+	}
+
+	b.Iteration.builds.Lock()
+	b.Iteration.builds.m[name] = build
+	b.Iteration.builds.Unlock()
+	return nil
+}
+
+func (b *Bucket) AddBuildArtifact(ctx context.Context, name string, partifacts ...PARtifact) error {
+	// NOOP
+	if b == nil {
+		return nil
+	}
+
+	build, ok := b.Iteration.builds.m[name]
+	if !ok {
+		return errors.New("no associated build found for the name " + name)
+	}
+
+	for _, artifact := range partifacts {
+		if build.CloudProvider == "" {
+			build.CloudProvider = artifact.ProviderName
+		}
+		build.PARtifacts = append(build.PARtifacts, artifact)
+	}
+
+	b.Iteration.builds.Lock()
+	b.Iteration.builds.m[name] = build
+	b.Iteration.builds.Unlock()
+
+	return nil
+}
+
+// Load defaults from environment variables
+func (b *Bucket) Canonicalize() {
```

Comment:
NIT: Should this be:
```suggestion
// Load defaults from environment variables
func (b *Bucket) LoadDefaultFromEnv() {
```
?